mycpp/marksweep_heap.h

```diff
@@ -0,0 +1,44 @@
+#ifndef MARKSWEEP_H
+#define MARKSWEEP_H
+
+#include <new>
+#include <unordered_set>
+
+typedef uint64_t u64;
+
+class MarkSweepHeap {
+ public:
+  MarkSweepHeap() {
+  }
+  void Init(int space_size);
+
+  void* Allocate(int);
+
+  void MarkAllReferences(Obj* obj);
+
+  void PushRoot(Obj** p) {
+    assert(roots_top_ < kMaxRoots);
+    roots_[roots_top_++] = p;
+  }
+
+  void PopRoot() {
+    roots_top_--;
+  }
+
+  void Collect(int to_space_size = 0);
+
+  int roots_top_;
+  Obj** roots_[kMaxRoots];  // These are pointers to Obj* pointers
+
+  u64 current_heap_bytes;
+  u64 max_heap_bytes;
+
+  // TODO(Jesse): This should really be in an 'internal' build
+  bool is_initialized_ =
+      true;  // mark/sweep heap doesn't need to be initialized
+
+  std::unordered_set<void*> AllAllocations;
```

Comment:
Hm do you think this is better than a linked list? (which is traditional)


I think it probably does simplify some things, e.g. having extra linked list node types

Reply:
The reason a linked-list is traditional is because people traditionally write their own allocators.  If you do that, the easiest method to keep track of free blocks is to use a linked list.

In _our_ case, `all_allocations_` could be either a hashtable or linked-list with minor tradeoffs in terms of memory and speed.

In _all_ cases, `marked_allocations_` _must_ be a hashtable (or some other structure with fast lookup) because we look up into it for every element in `all_allocations_`.

I made them both hashtables because we could see a (probably nominal) speedup if, at the end of collection, we clear `all_allocations_`, then swap the memory they point to.  Currently that swap is done in a loop at the end of the function.  I'm guessing that loop is going to be so fast (most of the time) it won't even show up on profiles, but decided I might as well leave the door open in case I'm wrong.